source/array.lua

```diff
@@ -0,0 +1,186 @@
+    Reimplemented by Matt Hargett, 2023
+
+]]
+--!strict
+local exports = {}
+export type Array<T> = { [number]: T }
+type Object = { [string]: any }
+exports.indexOf = function<T>(haystack: Array<T>, needle: T, startIndex_: number?): number
+  local length = #haystack
+  local startIndex = if startIndex_ == nil
+    then 1
+    else if startIndex_ < 1 then math.max(1, length - math.abs(startIndex_)) else startIndex_
+
+  for i = startIndex, length, 1 do
+    if haystack[i] == needle then
+      return i
+    end
+  end
+
+  -- we maintain the JS not found value, as it's often checked for explicitly and as < 0
+  return -1
+end
+
+-- we pulled a few refinement ideas from this stackoverflow article, but found:
+-- 1. no single one answer worked enough of the time in terms of transliterated JS expectations
+-- 2. most had very poor accuracy versus performance tradeoffs
+-- https://stackoverflow.com/questions/7526223/how-do-i-know-if-a-table-is-an-array/20958869#20958869
+exports.isArray = function(val: any): boolean
+  if type(val) ~= "table" then
+    return false
+  end
+
+  if next(val) == nil then
+    -- it's table with nothing in it, which we express is an array
+    -- this works 99% of the time for transliterated Lua
+    return true
+  end
+
+  local tableLength = #val
+  if tableLength == 0 then
+    -- getting past the preceding clause says the table isn't an empty iterable
+    -- if the length of the table is reported as 0, that means it has non-numeric indices
+    return false
+  end
+
+  -- the slow part, verifying each index is a positive, whole number. a Lua VM built-in would be nice.
+  for key, _ in pairs(val) do
+    if type(key) ~= "number" then
+      return false
+    end
+    if key < 1 then
+      -- Lua arrays start at 1, a 0 index means it's not a pure Lua array
+      return false
+    end
+    -- Lua TODO: would math.floor be faster? needs a benchmark
+    if (key % 1) ~= 0 then
+      -- if the number key isn't a whole number, it's not a pure Lua array
+      return false
+    end
+
+    if key > tableLength then
+      -- if we get a numeric key larger than the length operator reports, this isn't a contiguous array
+      return false
+    end
+
+    if key ~= tableLength then
+      -- if we're not at the end of a contiguous array, the value in the index slot should be non-nil
+      if nil == val[key + 1] then
+        return false
+      end
+    end
+  end
+
+  return true
+end
+
+-- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join
+exports.join = function<T>(array: Array<T>, separator: string?): string
+  -- this behavior is relied on by GraphQL and jest
+  return if 0 == #array
+    then ""
+    -- some lua implementations of concat don't behave when passed nil
+    else table.concat(array, separator or ",")
+end
+
+-- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
+exports.reverse = function(array: Array<any>): Array<any>
```

Comment:
This could be a generic function instead:
```suggestion
exports.reverse = function<T>(array: Array<T>): Array<T>
```